crates/subspace-fraud-proof/src/invalid_transaction_proof.rs

```diff
@@ -0,0 +1,373 @@
+    PClient: HeaderBackend<PBlock> + ProvideRuntimeApi<PBlock> + Send + Sync,
+    PClient::Api: ExecutorApi<PBlock, Hash>,
+    VerifierClient: VerifierApi,
+    DomainExtrinsicsBuilder: BuildDomainExtrinsics<PBlock>,
+    Exec: CodeExecutor + 'static,
+{
+    /// Constructs a new instance of [`InvalidTransactionProofVerifier`].
+    pub fn new(
+        primary_chain_client: Arc<PClient>,
+        executor: Arc<Exec>,
+        verifier_client: VerifierClient,
+        domain_extrinsics_builder: DomainExtrinsicsBuilder,
+    ) -> Self {
+        Self {
+            primary_chain_client,
+            executor,
+            verifier_client,
+            domain_extrinsics_builder,
+            _phantom: Default::default(),
+        }
+    }
+
+    fn fetch_domain_extrinsic(
+        &self,
+        domain_id: DomainId,
+        extrinsic_index: u32,
+        primary_hash: PBlock::Hash,
+        wasm_bundle: Vec<u8>,
+    ) -> Result<OpaqueExtrinsic, VerificationError> {
+        let domain_extrinsics = self
+            .domain_extrinsics_builder
+            .build_domain_extrinsics(domain_id, primary_hash, wasm_bundle)
+            .map_err(|_| VerificationError::FailedToBuildDomainExtrinsics)?;
+
+        let extrinsic = domain_extrinsics
+            .into_iter()
+            .nth(extrinsic_index as usize)
+            .ok_or(VerificationError::DomainExtrinsicNotFound(extrinsic_index))?;
+
+        Ok(OpaqueExtrinsic::from_bytes(&extrinsic)?)
+    }
+
+    fn fetch_primary_header(
+        &self,
+        domain_id: DomainId,
+        block_number: u32,
+    ) -> Result<PBlock::Header, VerificationError> {
+        let primary_hash: PBlock::Hash = self
+            .verifier_client
+            .primary_hash(domain_id, block_number)?
+            .into();
+
+        let header = self
+            .primary_chain_client
+            .header(primary_hash)?
+            .ok_or_else(|| {
+                sp_blockchain::Error::Backend(format!("Header for {primary_hash} not found"))
+            })?;
+
+        Ok(header)
+    }
+
+    /// Verifies the invalid transaction proof.
+    pub fn verify(
+        &self,
+        invalid_transaction_proof: &InvalidTransactionProof,
+    ) -> Result<(), VerificationError> {
+        let InvalidTransactionProof {
+            domain_id,
+            block_number,
+            domain_block_hash,
+            extrinsic_index,
+            storage_proof,
+        } = invalid_transaction_proof;
+
+        let header = self.fetch_primary_header(*domain_id, *block_number)?;
+        let primary_hash = header.hash();
+        let primary_parent_hash = *header.parent_hash();
+
+        let domain_runtime_code = retrieve_domain_runtime_code(
+            *domain_id,
+            primary_parent_hash,
+            &self.primary_chain_client,
+        )?;
+
+        let extrinsic = self.fetch_domain_extrinsic(
+            *domain_id,
+            *extrinsic_index,
+            primary_hash,
```

Comment:
The `InvalidTransactionProof::block_number` and `InvalidTransactionProof::domain_block_hash` is driven from `bundle_solution.proof_of_election().block_hash`, thus the `primary_hash` is referring to the primary block that the bundle built on top of instead of the primary block that contained the bundle, so we may get a wrong `extrinsic` here.